include/ensmallen_bits/lookahead/lookahead_impl.hpp

```diff
@@ -0,0 +1,191 @@
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_LOOKAHEAD_LOOKAHEAD_IMPL_HPP
+#define ENSMALLEN_LOOKAHEAD_LOOKAHEAD_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "lookahead.hpp"
+
+#include <ensmallen_bits/function.hpp>
+
+namespace ens {
+
+template<typename BaseOptimizerType, typename DecayPolicyType>
+inline Lookahead<BaseOptimizerType, DecayPolicyType>::Lookahead(
+    const BaseOptimizerType& baseOptimizer,
+    const double stepSize,
+    const size_t k,
+    const size_t maxIterations,
+    const double tolerance,
+    const DecayPolicyType& decayPolicy,
+    const bool exactObjective) :
+    baseOptimizer(baseOptimizer),
+    stepSize(stepSize),
+    k(k),
+    maxIterations(maxIterations),
+    tolerance(tolerance),
+    decayPolicy(decayPolicy),
+    exactObjective(exactObjective),
+    isInitialized(false)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename BaseOptimizerType, typename DecayPolicyType>
+template<typename DecomposableFunctionType,
+         typename MatType,
+         typename GradType,
+         typename... CallbackTypes>
+typename std::enable_if<IsArmaType<GradType>::value,
+typename MatType::elem_type>::type
+Lookahead<BaseOptimizerType, DecayPolicyType>::Optimize(
+    DecomposableFunctionType& function,
+    MatType& iterateIn,
+    CallbackTypes&&... callbacks)
+{
+  // Convenience typedefs.
+  typedef typename MatType::elem_type ElemType;
+  typedef typename MatTypeTraits<MatType>::BaseMatType BaseMatType;
+  typedef typename MatTypeTraits<GradType>::BaseMatType BaseGradType;
+
+  typedef Function<DecomposableFunctionType, BaseMatType, BaseGradType>
+      FullFunctionType;
+  FullFunctionType& f(static_cast<FullFunctionType&>(function));
+
+  // The decay policy internally use a templated class so that
+  // we can know MatType and GradType only when Optimize() is called.
+  typedef typename DecayPolicyType::template Policy<BaseMatType, BaseGradType>
+      InstDecayPolicyType;
+
+  // Make sure we have all the methods that we need.
+  traits::CheckDecomposableFunctionTypeAPI<FullFunctionType, BaseMatType,
+      BaseGradType>();
+  RequireFloatingPointType<BaseMatType>();
+  RequireFloatingPointType<BaseGradType>();
+  RequireSameInternalTypes<BaseMatType, BaseGradType>();
+
+  BaseMatType& iterate = (BaseMatType&) iterateIn;
+
+  // Check if the optimizer implements HasMaxIterations() and override the
+  // parameter with k.
+  if (callbacks::traits::HasMaxIterationsSignature<BaseOptimizerType>::value)
+  {
+    baseOptimizer.MaxIterations() = k;
+  }
+  else
+  {
+    Warn << "The BaseOptimizerType does not have a definition of "
+        << "MaxIterations(), don't override the number of iteration.";
```

Comment:
Maybe instead of `don't override the number of iteration`, it's more clear to say `the given value of k will be ignored` or something like this?  The user should probably be aware that in a situation like this, the inner optimizer will probably run until convergence and that will make this whole Lookahead strategy take way too long.  In fact, I wouldn't be opposed to forcing the existence of a `MaxIterations()` function and failing if it's not there.

Reply:
There are some optimizers that don't implement ``MaxIterations()`` like Differential Evolution which could work with the Lookahead optimizer without running forever, so unless you have a strong opinion here, I would go with the rewording.